### PR TITLE
[JENKINS-53346] fix for same env vars for different branches

### PR DESCRIPTION
**Pull request by tmcsantos**
## [JENKINS-53346](https://issues.jenkins-ci.org/browse/JENKINS-53346) - attempting to fix checkout returning same values on second different call

The problem relies on the inability of GitSCM.getBuildData returning the right BuildData for the instance.  It was only comparing the urls, ignoring the branch name.  So in the end, a different branch name would get the first registered checkout's BuildData.

This change will make it search not only for the url but also for the branch name.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [ ] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Further comments

~Although this seems to fix the mentioned issue, the changes causes other tests to start failing.  I did not have the time to look at them!~

@MarkEWaite could you please review my approach?

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -1767,22 +1766,17 @@ public BuildData copyBuildData(Run build) {
...
      */
     public @CheckForNull BuildData getBuildData(Run build) {
-        BuildData buildData = null;
         while (build != null) {
             List<BuildData> buildDataList = build.getActions(BuildData.class);
             for (BuildData bd : buildDataList) {
                 if (bd != null && isRelevantBuildData(bd)) {
-                    buildData = bd;
-                    break;
+                    return bd;
```
Comment (MarkEWaite) [inline src/main/java/hudson/plugins/git/GitSCM.java]: This changes the semantics of the `getBuildData` call.  Previously it return the last recorded build data, now it returns the first.

Is that intentional?  How does that help with the problem?

**Comment by tmcsantos** [comment]
After taking a better look at the remaining tests that failed, I concluded that the initial approach was very incomplete. Updated the PR to complement with the edge cases encountered.

On the other end, this one seemed to have also fixed [JENKINS-44762](https://issues.jenkins-ci.org/browse/JENKINS-44762).  Guess they were somewhat related.

@MarkEWaite could you please review
